core/chains/evm/gas/rollups/op_l1_oracle.go

```diff
@@ -0,0 +1,440 @@
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+	"github.com/smartcontractkit/chainlink-common/pkg/utils"
+
+	gethtypes "github.com/ethereum/go-ethereum/core/types"
+
+	"github.com/smartcontractkit/chainlink/v2/common/client"
+	"github.com/smartcontractkit/chainlink/v2/common/config"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/assets"
+	evmclient "github.com/smartcontractkit/chainlink/v2/core/chains/evm/client"
+)
+
+type OptimismL1Oracle interface {
+	L1Oracle
+	GetDAGasPrice(ctx context.Context) (*big.Int, error)
+}
+
+// Reads L2-specific precompiles and caches the l1GasPrice set by the L2.
+type optimismL1Oracle struct {
+	services.StateMachine
+	client     l1OracleClient
+	pollPeriod time.Duration
+	logger     logger.SugaredLogger
+	chainType  config.ChainType
+
+	l1OracleAddress     string
+	gasPriceMethod      string
+	l1GasPriceMethodAbi abi.ABI
+	l1GasPriceMu        sync.RWMutex
+	l1GasPrice          priceEntry
+
+	gasCostMethod      string
+	l1GasCostMethodAbi abi.ABI
+
+	chInitialised chan struct{}
+	chStop        services.StopChan
+	chDone        chan struct{}
+
+	isEcotoneMethodAbi abi.ABI
+
+	l1BaseFeeCalldata    []byte
+	isEcotoneCalldata    []byte
+	getL1GasUsedCalldata []byte
+	getL1FeeCalldata     []byte
+
+	isEcotone        bool
+	isEcotoneCheckTs int64
+}
+
+const (
+	// OPStackGasOracle_isEcotone fetches if the OP Stack GasPriceOracle contract has upgraded to Ecotone
+	OPStackGasOracle_isEcotone = "isEcotone"
+	// OPStackGasOracle_getL1GasUsed fetches the l1 gas used for given tx bytes
+	OPStackGasOracle_getL1GasUsed = "getL1GasUsed"
+	// OPStackGasOracle_isEcotonePollingPeriod is the interval to poll if chain has upgraded to Ecotone
+	// Set to poll every 4 hours
+	OPStackGasOracle_isEcotonePollingPeriod = 14400
+	// OPStackGasOracleAddress is the address of the precompiled contract that exists on OP stack chain.
+	// OPStackGasOracle_l1BaseFee fetches the l1 base fee set in the OP Stack GasPriceOracle contract
+	// OPStackGasOracle_l1BaseFee is a hex encoded call to:
+	// `function l1BaseFee() external view returns (uint256);`
+	OPStackGasOracle_l1BaseFee = "l1BaseFee"
+	// OPStackGasOracle_getL1Fee fetches the l1 fee for given tx bytes
+	// OPStackGasOracle_getL1Fee is a hex encoded call to:
+	// `function getL1Fee(bytes) external view returns (uint256);`
+	OPStackGasOracle_getL1Fee = "getL1Fee"
+	// This is the case for Optimism and Base.
+	OPGasOracleAddress = "0x420000000000000000000000000000000000000F"
+	// GasOracleAddress is the address of the precompiled contract that exists on Kroma chain.
+	// This is the case for Kroma.
+	KromaGasOracleAddress = "0x4200000000000000000000000000000000000005"
+	// ScrollGasOracleAddress is the address of the precompiled contract that exists on Scroll chain.
+	ScrollGasOracleAddress = "0x5300000000000000000000000000000000000002"
+)
+
+func NewOpStackL1GasOracle(lggr logger.Logger, ethClient l1OracleClient, chainType config.ChainType) OptimismL1Oracle {
```

Comment:
To go with the previous comment, that means we can just directly return:
```suggestion
func NewOpStackL1GasOracle(lggr logger.Logger, ethClient l1OracleClient, chainType config.ChainType) L1Oracle {
```